remotes/docker/resolver.go

```diff
@@ -382,37 +374,40 @@ func (r *dockerResolver) Resolve(ctx context.Context, ref string) (string, ocisp
 }
 
 func (r *dockerResolver) Fetcher(ctx context.Context, ref string) (remotes.Fetcher, error) {
-	refspec, err := reference.Parse(ref)
+	base, err := r.resolveDockerBase(ref)
 	if err != nil {
 		return nil, err
 	}
 
-	base, err := r.base(refspec)
+	return dockerFetcher{
+		dockerBase: base,
+	}, nil
+}
+
+func (r *dockerResolver) Pusher(ctx context.Context, ref string) (remotes.Pusher, error) {
+	base, err := r.resolveDockerBase(ref)
 	if err != nil {
 		return nil, err
 	}
 
-	return dockerFetcher{
+	return dockerPusher{
 		dockerBase: base,
+		object:     base.refspec.Object,
+		tracker:    r.tracker,
 	}, nil
 }
 
-func (r *dockerResolver) Pusher(ctx context.Context, ref string) (remotes.Pusher, error) {
+func (r *dockerResolver) resolveDockerBase(ref string) (*dockerBase, error) {
 	refspec, err := reference.Parse(ref)
 	if err != nil {
 		return nil, err
 	}
 
-	base, err := r.base(refspec)
-	if err != nil {
-		return nil, err
+	if refspec.Object == "" {
```

Comment:
The `Object` is allowed to be empty for Fetch and Push, only Resolve requires the object to be provided. This check can just be moved there

Reply:
I am not sure about it but removed.